pkg/registry/memory/ns_server.go

```diff
@@ -13,43 +29,43 @@ import (
 )
 
 type networkServiceRegistryServer struct {
-	networkServices     NetworkServiceSyncMap
-	executor            serialize.Executor
-	eventChannels       []chan *registry.NetworkService
-	eventChannelSize    int
-	eventChannelsLocker sync.Mutex
+	networkServices  map[string]*registry.NetworkService
+	executor         serialize.Executor
+	eventChannels    []chan *registry.NetworkService
+	eventChannelSize int
 }
 
 func (n *networkServiceRegistryServer) Register(ctx context.Context, ns *registry.NetworkService) (*registry.NetworkService, error) {
-	n.networkServices.Store(ns.Name, ns)
 	n.executor.AsyncExec(func() {
-		n.eventChannelsLocker.Lock()
+		n.networkServices[ns.Name] = ns
 		for _, ch := range n.eventChannels {
 			ch <- ns
 		}
-		n.eventChannelsLocker.Unlock()
 	})
 	return next.NetworkServiceRegistryServer(ctx).Register(ctx, ns)
```

Comment:
We need to cleanup if next return error.

Reply:
nice catch, fixed